### PR TITLE
MINOR: disable JmxTool in kafkatest console-consumer by default

### PR DESCRIPTION
**Pull request by brianbushree**
### what
Provide the ability to optionally enable the JmxTool inside the kafkatest console-consumer

### why
For tests that don't use these JMX metrics, we should allow users to disable this metrics collection.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: tests/kafkatest/services/monitor/jmx.py [hunk 1]
```diff
@@ -27,7 +27,8 @@ class JmxMixin(object):
     - we assume the service using JmxMixin also uses KafkaPathResolverMixin
     - this uses the --wait option for JmxTool, so the list of object names must be explicit; no patterns are permitted
     """
-    def __init__(self, num_nodes, jmx_object_names=None, jmx_attributes=None, jmx_poll_ms=1000, root="/mnt"):
+    def __init__(self, num_nodes, jmx_enabled=True, jmx_object_names=None, jmx_attributes=None, jmx_poll_ms=1000, root="/mnt"):
```
Comment (hachikuji) [inline tests/kafkatest/services/monitor/jmx.py]: I wonder if it would be better to rely on the `JmxTool` wrapper that @mumrah added here: https://github.com/apache/kafka/blob/trunk/tests/kafkatest/tests/core/fetch_from_follower_test.py#L31. We can pull this up into a more convenient location and change console consumer to only construct an instance if `jmx_enabled` is set. That seems a little cleaner than supporting an option to turn off the mixin.

Comment (brianbushree) [reply tests/kafkatest/services/monitor/jmx.py]: makes sense!

**Comment by brianbushree** [comment]
@hachikuji can you take another look please? 

is this what you were thinking?

File: tests/kafkatest/services/monitor/jmx.py [hunk 2]
```diff
@@ -41,10 +42,9 @@ def __init__(self, num_nodes, jmx_object_names=None, jmx_attributes=None, jmx_po
         self.jmx_tool_log = os.path.join(root, "jmx_tool.log")
         self.jmx_tool_err_log = os.path.join(root, "jmx_tool.err.log")
 
-    def clean_node(self, node):
+    def clean_node(self, idx, node):
```
Comment (brianbushree) [inline tests/kafkatest/services/monitor/jmx.py]: maybe instead we can do this?

```
def clean_node(self, node, idx=None):
```

then if `idx` is specified we know not to call `self.idx(node)`?

File: tests/kafkatest/services/console_consumer.py [hunk 3]
```diff
@@ -102,6 +101,11 @@ def __init__(self, context, num_nodes, kafka, topic, group_id="test-consumer-gro
         for node in self.nodes:
             node.version = version
 
+        self.jmx_enabled = jmx_enabled
```
Comment (hachikuji) [inline tests/kafkatest/services/console_consumer.py]: There's a little redundancy in having both `jmx_enabled` and `jmx_tool` fields. Could we use `jmx_enabled` just to construct `jmx_tool`? After that, then we could replace `if self.jmx_enabled` with `if self.jmx_tool`. What do you think?

File: tests/kafkatest/services/console_consumer.py [hunk 4]
```diff
@@ -59,7 +59,8 @@ class ConsoleConsumer(KafkaPathResolverMixin, JmxMixin, BackgroundThreadService)
 
     def __init__(self, context, num_nodes, kafka, topic, group_id="test-consumer-group", new_consumer=True,
                  message_validator=None, from_beginning=True, consumer_timeout_ms=None, version=DEV_BRANCH,
-                 client_id="console-consumer", print_key=False, jmx_object_names=None, jmx_attributes=None,
+                 client_id="console-consumer", print_key=False,
+                 jmx_enabled=True, jmx_object_names=None, jmx_attributes=None,
```
Comment (hachikuji) [inline tests/kafkatest/services/console_consumer.py]: Can we infer `jmx_enabled` based on `jmx_object_names` and `jmx_attributes`? Are there any cases we would want the jmx tool if these are both `None`?

**Review by hachikuji — COMMENTED** [review]
Thanks, left a couple minor suggestions.

File: tests/kafkatest/services/console_consumer.py (hunk 4 above)
Comment (brianbushree) [reply tests/kafkatest/services/console_consumer.py]: looks like we already refuse to start if not objects are specified

```python
    def start_jmx_tool(self, idx, node):
        if self.jmx_object_names is None:
            self.logger.debug("%s: Not starting jmx tool because no jmx objects are defined" % node.account)
            return
        ...
```

so sounds like we can just say
```python
    self.jmx_enabled = False if self.jmx_object_names is None else True
```

File: tests/kafkatest/services/console_consumer.py [hunk 5]
```diff
@@ -161,8 +162,10 @@ def start_cmd(self, node):
         else:
             args['kafka_opts'] = self.security_config.kafka_opts
 
-        cmd = "export JMX_PORT=%(jmx_port)s; " \
-              "export LOG_DIR=%(log_dir)s; " \
+        cmd = ""
+        if self.jmx_tool is not None:
```
Comment (hachikuji) [inline tests/kafkatest/services/console_consumer.py]: nit: these can be replaced by `if self.jmx_tool`

File: tests/kafkatest/services/console_consumer.py [hunk 6]
```diff
@@ -102,6 +100,10 @@ def __init__(self, context, num_nodes, kafka, topic, group_id="test-consumer-gro
         for node in self.nodes:
             node.version = version
 
+        self.jmx_tool = JmxTool(context, jmx_object_names=jmx_object_names, jmx_attributes=(jmx_attributes or []),
+                                root=ConsoleConsumer.PERSISTENT_ROOT) \
+                        if jmx_object_names is not None else None
```
Comment (hachikuji) [inline tests/kafkatest/services/console_consumer.py]: Should we check `jmx_attributes` as well?

**Review by hachikuji — COMMENTED** [review]
@brianbushree Thanks, left a couple minor comments. Just to confirm, have you verified system tests are still passing with these changes? No need to check them all, maybe just make sure some test case which uses `has_partitions_assigned` (for example) still works.

File: tests/kafkatest/services/console_consumer.py (hunk 6 above)
Comment (brianbushree) [reply tests/kafkatest/services/console_consumer.py]: I believe we only want to check `jmx_object_names` since it seems like it's the only required parameter for this JmxTool wrapper.

Comment (brianbushree) [reply tests/kafkatest/services/console_consumer.py]: correction: both are required here. we should be checking both

Comment (mumrah) [reply tests/kafkatest/services/console_consumer.py]: This syntax is a bit hard to follow with the conditional at the end. Can you rewrite it to something like:

```python
self.jmx_tool = None
if jmx_object_names is not None:
    self.jmx_tool = ...
```

(and also check the attributes as mentioned above)

File: tests/kafkatest/services/console_consumer.py [hunk 7]
```diff
@@ -285,7 +290,8 @@ def clean_node(self, node):
         if self.alive(node):
             self.logger.warn("%s %s was still alive at cleanup time. Killing forcefully..." %
                              (self.__class__.__name__, node.account))
-        JmxMixin.clean_node(self, node)
+        idx = self.idx(node)
+        self.jmx_tool.clean_node(idx, node)
```
Comment (mumrah) [inline tests/kafkatest/services/console_consumer.py]: Need to check if `self.jmx_tool` is None here?

**Comment by brianbushree** [comment]
looks like I overlooked implementing proper default values with regard to inferring `jmx_enabled` from `jmx_object_names`/`jmx_attributes`...

currently we initialize the `jmx_object_names` and `jmx_attributes` here
```python
self.jmx_object_names = ["kafka.consumer:type=consumer-coordinator-metrics,client-id=%s" % self.client_id]
self.jmx_attributes = ["assigned-partitions"]
```
if `jmx_object_names` is none (guessing we do this because the default value depends on another parameter, `client-id`)

unfortunately this means using `jmx_object_names=None` implies that we should use the default value... 

seems like we have two options:
1. stick with a simple `jmx_enabled` flag
2. change the behavior of `jmx_object_names` such that we can pass in a string format and always replace `%s` with the client-id

@hachikuji what do you think? did I miss any other options?

File: tests/kafkatest/services/console_consumer.py (hunk 7 above)
Comment (brianbushree) [reply tests/kafkatest/services/console_consumer.py]: good catch! thanks

Comment (mumrah) [reply tests/kafkatest/services/console_consumer.py]: Actually there are a few places like this. Maybe we can figure out how to let jmx_tool always be set even if we're not scraping any JMX objects? I think that's how the mixin worked before.

Comment (brianbushree) [reply tests/kafkatest/services/console_consumer.py]: ya, that makes sense. the less special-casing the better

**Comment by hachikuji** [comment]
I think the only reason we use the jmx tool by default is for the purpose of `ConsoleConsumer.has_partitions_assigned`. The only usage of this as far as I can tell is in `produce_consume_validate.py` and does not seem to be contributing significantly. I am thinking we can just remove this function and let the jmx tool be disabled by default. What do you think?

**Comment by brianbushree** [comment]
Ya I agree completely. It makes sense to disable this by default since it seems there are only a few usages and makes this PR much more complicated. I'll remove this function and some of the associated logic.

File: tests/kafkatest/services/console_consumer.py [hunk 8]
```diff
@@ -102,6 +100,10 @@ def __init__(self, context, num_nodes, kafka, topic, group_id="test-consumer-gro
         for node in self.nodes:
             node.version = version
 
+        self.jmx_tool = JmxTool(context, num_nodes=num_nodes,
+                                jmx_object_names=jmx_object_names, jmx_attributes=(jmx_attributes or []),
+                                root=ConsoleConsumer.PERSISTENT_ROOT)
```
Comment (brianbushree) [inline tests/kafkatest/services/console_consumer.py]: @mumrah to your earlier point, I've tried to avoid special casing by always initializing a `JmxTool` regardless of whether it'll be used

@hachikuji however, now that this is in place, it seems like we should maybe just stick with mixin approach since we're always initializing this object anyways? (this way we can also avoid some of the `clean_node` changes required here)

with that being said, I do like the `consumer.jmx_tool` namespace isolation

Comment (hachikuji) [reply tests/kafkatest/services/console_consumer.py]: @brianbushree I don't mind going back to the mixin if you think that's better. I'll leave it up to you.

Comment (brianbushree) [reply tests/kafkatest/services/console_consumer.py]: I think this mixin approach is better for two reasons here:
1. it should minimize this PR
2. seems less like a one-off approach since we already use the mixin everywhere

I'm going to refactor this PR accordingly.

**Comment by brianbushree** [comment]
These are the tests that seem to rely on this jmx functionality in the consumer:
- `fetch_from_follower_test.py`
- `streams_simple_benchmark_test.py`
- `quota_test.py`

I've tested these locally and they are passing (I also tested `compression_test.py` which does not rely on this jmx functionality)

**Review by hachikuji — APPROVED** [review]
LGTM.
